<script src="../lib/fx.js"></script>

<!--
    1. async:await
-->
<script>
  function delay(a, t) {
    return new Promise((resolve) => setTimeout(() => resolve(a), t));
  }

  async function f1() {
    const a = await delay(10, 500);
    const b = await delay(5, 500);

    return a + b;
  }

  //   log(f1());
  //   f1().then(log);
  //   go(f1(), log);
  //   (async () => {
  //     log(await f1());
  //   })();
</script>

<!--
    2. [QnA] Array.prototype.map이 있는데 왜 FxJS의 map 함수가 필요한지?
-->
<script>
  async function f2() {
    const list = [1, 2, 3, 4];
    const res = await list.map(async (a) => await delay(a * a, 100));
    // log(res);
  }
  f2();

  async function f3() {
    const list = [1, 2, 3, 4];
    const res = await map(async (a) => await delay(a * a, 100), list);
    // log(res);
  }
  f3();

  const f4 = () => map(async (a) => await delay(a * a, 100), [1, 2, 3, 4]);

  //   log(f4());
  (async () => {
    // log(await f4());
  })();
</script>

<!--
    3. [QnA] 이제 비동기는 async/await으로 제어할 수 있는데 왜 파이프라인이 필요한지?
    - async/await 는 문장형으로 비동기를 다루기 위해 등장한 것이다.
    - 파이프라인은 명령형으로 비동기를 처리하기 위해 등장한 것이다.
    => 비동기에 대한 처리가 목적이 아닌, 서로 다른 문제에 대해 해결하고자한 해결책이다.
-->
<script>
  const delay100 = (a) => delay(a, 100);

  let list = range(8);

  // 파이프라인으로 작성한 함수
  function f5(list) {
    return go(
      list,
      L.map((a) => delay100(a * a)),
      L.filter((a) => delay100(a % 2)),
      L.map((a) => delay100(a + 1)),
      take(3),
      reduce((a, b) => delay100(a + b))
    );
  }
  //   go(f5(list), (a) => log(a, "f5"));

  // async/await으로 작성한 함수
  async function f6(list) {
    let temp = [];
    for (const a of list) {
      const b = await delay100(a * a);
      if (await delay100(b % 2)) {
        const c = await delay100(b + 1);
        temp.push(c);
        if (temp.length === 3) break;
      }
    }
    let res = 0;
    for (const a of temp) {
      res = await delay100(res + a);
    }
    return res;
  }
  //   go(f6(list), (a) => log(a, "f6"));
</script>

<!--
    4. [QnA] async/await과 파이프라인을 같이 사용하기도 하나요?
-->
<script>
  async function f52(list) {
    const r1 = await go(
      list,
      L.map((a) => delay100(a * a)),
      L.filter((a) => delay100(a % 2)),
      L.map((a) => delay100(a + 1)),
      take(3),
      reduce((a, b) => delay100(a + b))
    );

    const r2 = await go(
      list,
      L.map((a) => delay100(a * a)),
      L.filter((a) => delay100(a % 2)),
      reduce((a, b) => delay100(a + b))
    );

    const r3 = await delay100(r1 + r2);

    return r3 + 10;
  }
  //   go(f52(list), (a) => log(a, "f52"));
</script>

<!--
    5. [QnA] 동기 상황에서 에러 핸들링은 어떻게 해야하는지?
-->
<script>
  function f7(list) {
    try {
      return list
        .map((a) => JSON.parse(a))
        .filter((a) => a % 2)
        .slice(0, 2);
    } catch (e) {
      log(e);
      return [];
    }
  }
  //   f7(list);
</script>

<!--
    6. [QnA] 비동기 상황에서 에러 핸들링은 어떻게 해야하는지?
-->
<script>
  async function f8(list) {
    try {
      return await list
        .map(
          async (a) => await new Promise((resolve) => resolve(JSON.parse(a)))
        )
        .filter((a) => a % 2)
        .slice(0, 2);
    } catch (e) {
      log(e);
    }
  }
  //   f8(["0", "1", "2", "{"])
  //     .then(log)
  //     .catch((e) => {
  //       log(e);
  //     });
</script>

<!--
    6. [QnA] 동기/비동기 에러 핸들링에서의 파이프라인의 이점은?
-->
<script>
  async function f9(list) {
    try {
      return await go(
        list,
        map(
          async (a) => await new Promise((resolve) => resolve(JSON.parse(a)))
        ),
        filter((a) => a % 2),
        take(2)
      );
    } catch (e) {
      log("에러 핸들링 확인", e);
      return [];
    }
  }
  f9(["0", "1", "2", "{"])
    .then(log)
    .catch((e) => {
      log("외부 에러 핸들링", e);
    });
</script>
